sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java

```diff
@@ -315,17 +331,25 @@ public Objects listObjects(String bucket, String prefix, @Nullable String pageTo
    * if the resource does not exist.
    */
   @VisibleForTesting
-  List<Long> fileSizes(Collection<GcsPath> paths) throws IOException {
-    List<StorageObject[]> results = Lists.newArrayList();
-    executeBatches(makeGetBatches(paths, results));
+  List<Long> fileSizes(List<GcsPath> paths) throws IOException {
+    List<StorageObjectOrIOException> results = getObjects(paths);
 
     ImmutableList.Builder<Long> ret = ImmutableList.builder();
-    for (StorageObject[] result : results) {
-      ret.add(result[0].getSize().longValue());
+    for (StorageObjectOrIOException result : results) {
+      ret.add(toFileSize(result));
     }
     return ret.build();
   }
 
+  private Long toFileSize(StorageObjectOrIOException storageObjectOrIOException)
+      throws IOException {
+    if (storageObjectOrIOException.ioException() != null) {
+      throw storageObjectOrIOException.ioException();
+    }
+    return checkNotNull(storageObjectOrIOException.storageObject(), "storageObject")
```

Comment:
This should never happen, by construction - it should be impossible to receive a nonnull `StorageObjectOrIOException` with two `null` fields.